packages/teams-js/src/internal/communication.ts

```diff
@@ -257,6 +264,90 @@ export function sendMessageToParentAsync<T>(
   });
 }
 
+type SimpleType = string | number | boolean | null | undefined | SimpleType[];
+
+function isSerializable(arg: unknown): arg is ISerializable {
+  return (
+    arg !== undefined &&
+    arg !== null &&
+    (arg as ISerializable).serialize !== undefined &&
+    typeof (arg as ISerializable).serialize === 'function'
+  );
+}
+
+function serializeItemArray(items: (SimpleType | ISerializable)[]): unknown[] {
+  return items.map((item) => {
+    if (isSerializable(item)) {
+      return item.serialize();
+    } else {
+      return item;
+    }
+  });
+}
+
+/**
+ * Call a function in the host and receive a response. If the host returns an {@link SdkError} instead of a normal response, this function will throw a new Error containing the SdkError's information
+ *
+ * @param apiVersionTag A unique tag used to identify the API version for telemetry purposes.
+ * @param functionName The function name to call in the host.
+ * @param responseHandler When the host responds, this handler will validate and deserialize the response.
+ * @param args A collection of data to pass to the host. This data must be an array of either simple types or objects that implement {@link ISerializable}.
+ * @param errorChecker This optional property can be used to override the default ErrorChecking this function uses to decide whether to throw the host response as a new Error. Specify this is your function needs to do any logic verifying that the object received is an error that goes beyond the logic found in {@link isSdkError}.
+ *
+ * @returns The response received from the host after deserialization.
+ *
+ * @throws An Error containing the SdkError information ({@link SdkError.errorCode} and {@link SdkError.message}) if the host returns an SdkError, or an Error if the response from the host is an unexpected format.
+ */
+export async function callFunctionInHostAndHandleResponse<ReceivedFromHost, DeserializedFromHost>(
+  apiVersionTag: string,
+  functionName: string,
+  responseHandler: ResponseHandler<ReceivedFromHost, DeserializedFromHost>,
+  args: (SimpleType | ISerializable)[],
+  errorChecker?: (response: unknown) => response is SdkError,
+): Promise<DeserializedFromHost> {
+  const argsSafeToTransfer = serializeItemArray(args);
+  const [response] = await sendMessageToParentAsync<[ReceivedFromHost | SdkError]>(
+    apiVersionTag,
+    functionName,
+    argsSafeToTransfer,
+  );
+
+  if ((errorChecker && errorChecker(response)) || (!errorChecker && isSdkError(response))) {
+    throw new Error(`${response.errorCode}, message: ${response.message ?? 'None'}`);
+  } else if (!responseHandler.validate(response as ReceivedFromHost)) {
+    throw new Error(`${ErrorCode.INTERNAL_ERROR}, message: Invalid response from host`);
+  } else {
+    return responseHandler.deserialize(response as ReceivedFromHost);
```

Comment:
>deserialize [](http://example.com/codeflow?start=27&length=11)

I have been thinking about this and it seems like there's some semantic asymmetry in how teamjs sends and receives data for this communication.

Is the intended design that the hub SDK can send a single data value to teamjs, or that it can send multiple data values?

The current design suggests that teamsjs is allowed to send multiple data values to the hub SDK (`args` is an array where each element can individually be a `SimpleType` or a type that serializes) but that the hub SDK can only send a single value to teamsjs, and that single value is deserialized as a whole. If you are in the hub SDK and want to send multiple values to teamsjs, you have to either send them as an object or send them in an array (which I guess is then wrapped in its own array?).

Such a design isn't necessarily too weird, if you think about how you would just regularly call a function, functions can have multiple input parameters and can only have a single return value. If you want to make a function return multiple values, you have to make the return type be either an object or an array or a tuple.

I wanted to make sure I understood the intended design, as to be observed by capability developers. Should they think about the hub SDK being able to send back multiple values, or only a single value (and if they want multiples they have to wrap them in an object or array)?

Reply:
You're correct that there's asymmetry there, and it's by design for the reasons you're guessing. In order to make these functions as semantically simple as I can (because they are inherently hard to understand) I wanted to make them mimic how developers think of calling "normal" functions: pass the function a list of arguments, receive a single return value. That return value could be an object, a simple type, or an array of things, but whatever it is it's _just one thing_.

Can someone do something fancy and return a list of various return values from the host? Sure. Is it more complicated to understand? Yes. Does it actually help with anything? No!